<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>SeqPlots by Przemol</title>
    
    <link rel="shortcut icon" href="favicon.ico">
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>

        <p>
          Install <b> SeqPlots Bioconductor package</b> :
          <a href="http://www.bioconductor.org/packages/release/bioc/html/seqplots.html">
            <img src="http://www.bioconductor.org/images/logo_bioconductor.gif" border="0" class="masthead_logo" height="78" width="260" alt="Bioconductor - open source software for bioinformatics">
          </a>
        </p>

        Get <b>SeqPlots Mac OS X boundle</b>:
        <ul style="width:90px;">
          <li><a href="https://github.com/Przemol/seqplots_osx/zipball/master">Download <strong>OS X app</strong></a></li>
        </ul>


        <p class="view"><a href="https://github.com/Przemol/seqplots">View the Main Project on GitHub <small>przemol/seqplots</small></a></p>
        <p class="docs"><a href="help.html">Read User Guide <small><a href="SeqPlots.pdf">...also as PDF file</a></small></a></p>
        <p class="visit"><a href="https://github.com/Przemol/seqplots/wiki">Visit Wiki Pages <small>przemol/seqplots/wiki</small></a></p>


      </header>
      <section>
        <h1>
<a id="" class="anchor" href="#" aria-hidden="true"><span class="octicon octicon-link"></span></a><img src="https://github.com/Przemol/seqplots/wiki/img/LogoSeqPlotsLabeled.png" alt="SeqPlots">
</h1>

<blockquote>
<p>An interactive tool for visualizing track signals and sequence motif densities along genomic features using average plots and heatmaps</p>
</blockquote>

<p>
  <img src="img/SeqPlots_fig1_web.png" alt="Examples of Seq Plots interface and outputs">
</p>

<h2>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h2>

<p>The SeqPlots is universal, web browser based tool for plotting
average track signal (e.g. reads coverage) and sequence motif density over
user specified genomic features. The data are visualized on linear plot,
with error estimates as fields, or as series of heatmaps that can be sorted
and clustered. The dual-purpose implementation allows running the software
locally on desktop or deploying it on server. The unique features of our
software are collaboration and data sharing capabilities, as well as
ability to store pre-calculated result matrixes, that combine many
sequencing experiments and in-silico generated tracks with multiple
different features. These binaries can be further used to generate new
combination plots on fly, run automated batch operations or share with
colleagues, who can adjust their plotting parameters without loading actual
tracks and recalculating numeric values.</p>

<hr />

<h2>
<a id="mac-os-x-bundle-app" class="anchor" href="#mac-os-x-bundle-app" aria-hidden="true"><span class="octicon octicon-link"></span></a>Mac OS X bundle application</h2>

<p>Mac OS X bundle is an easy way to use SeqPlots for Mac OS X users. It contains R binaries pre-installed Bioconductor dependencies. Additionally, the reference genome for <em>Caenorhabditis elegans</em> is included. The sequences for other popular model organisms can be downloaded using graphical user interface.</p>

<h3>How to install</h3>
<ol>
<li><p>Download the compressed application from
<a href="https://github.com/Przemol/seqplots_osx/zipball/master">https://github.com/Przemol/seqplots_osx/zipball/master</a></p>
</li>
<li><p>Double-click on downloaded file to unzip the contents of archive</p>
</li>
<li><p>Drag and drop SeqPlots.app to Applications folder</p>
</li>
</ol>

<h3>How to start</h3>
<p>
  Please refer to <a href="http://przemol.github.io/seqplots/help.html#Howtostart">user guide</a>.


</p>
<hr />


<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Bioconductor package in R</h2>

<p>
  The Bioconductor package runs SeqPlots locally on desktop computers from R environment. It is available for Windows, GNU/Linux and Mac OS X operating systems.
</p>

<h3>How to install</h3>

<ol>

<li><p> Install R from <a href="http://cran.rstudio.com/">here</a>.</p> </li>

<li><p> To install SeqPlots package, start R and enter:</p>

<div class="highlight highlight-r"><pre>source(<span class="pl-s"><span class="pl-pds">"</span>http://bioconductor.org/biocLite.R<span class="pl-pds">"</span></span>)
biocLite(<span class="pl-s"><span class="pl-pds">"</span>seqplots<span class="pl-pds">"</span></span>)</pre></div>
</li>
</ol>

<h3>How to start</h3>
<p>
  Please refer to
  <a href="http://przemol.github.io/seqplots/help.html#Howtostart12">user guide</a> or
  <a href="http://www.bioconductor.org/packages/release/bioc/html/seqplots.html">Bioconductor webpage</a>.
</p>
<hr />

<h2>
<a id="quick-start" class="anchor" href="#quick-start" aria-hidden="true"><span class="octicon octicon-link"></span></a>Start using SeqPlots</h2>

<p>
  SeqPlots is easy to start with thanks to intuitive, web browser based graphical user interface.
  The good place to learn about SeqPlots' workflow is the <a href="http://przemol.github.io/seqplots/help.html#Quickstartguide"><b>Quick start guide</b></a>.
</p>
<p>
  <a href="http://przemol.github.io/seqplots/help.html#Quickstartguide"><b>SeqPlots user's guide</b></a> explains all options and controls, as well as more covers advanced topics, like running SeqPlots as server application.
  Help about specific SeqPlots controls is available from within the web application, after clicking on orange (?) button.
</p>






        <h2>
<a id="live-demo" class="anchor" href="#live-demo" aria-hidden="true"><span class="octicon octicon-link"></span></a>Live demo</h2>

<p>The working demonstration preview of SeqPlots is available on ShinyApps.io:
<strong><a href="https://seqplots.shinyapps.io/seqplots/">https://seqplots.shinyapps.io/seqplots/</a></strong></p>
<small>
<p>Demo limits:</p>

<ul>
<li>  the application contains some pre-loaded data (<em>C. elegans</em> genes and histone modifications)</li>
<li>  size of uploaded files is limited to 30MB, please bin or subset your tracks to fit within the limit (due to ShinyApps beta upload limitations)</li>
<li>  only <em>C. elegans</em> genome is available (due ShinyApps beta space limits)</li>
<li>  uploaded files will be deleted after some time</li>
</ul>
</small>

      </section>
      <footer>

        <p><strong>Live demo</strong><small> - a demonstration preview of SeqPlots on ShinyApps.io:</small>
          <strong><a href="https://seqplots.shinyapps.io/seqplots/">https://seqplots.shinyapps.io/seqplots/</a></strong>
        <br/><small>See <a href="#live-demo">here</a> for demo limitations.</small>
        </p>

        <p>Project status: <br/>
          <a href="https://travis-ci.org/Przemol/seqplots"><img src="https://travis-ci.org/Przemol/seqplots.svg?branch=master" alt="Build Status"></a>
        </p>

        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-48467041-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
